controller/controller.go

```diff
@@ -232,6 +232,52 @@ func (c *notificationController) processApp(app *unstructured.Unstructured, logE
 	return nil
 }
 
+// Checks if the application SyncStatus has been refreshed by Argo CD after an operation has completed
+func (c *notificationController) isAppSyncStatusRefreshed(app *unstructured.Unstructured, logEntry *log.Entry) bool {
```

Comment:
Any chance we have this check inside of the trigger [condition](https://github.com/argoproj-labs/argocd-notifications/blob/master/builtin/triggers.go#L29)? 

The condition has access to the  `time.Parse` function.  This should be enough to verify that `reconciledAt ` is after sync completion. E.g.

```
app.status.operationState.phase in ['Succeeded'] && time.Parse(app.status.operationState.finishedAt).Before(time.Parse(app.status.reconciledAt))
```

The condition might fail and it is treated as false, so it is ok if `finishedAt` or `reconciledAt` are empty and `time.Parse` fails.

Reply:
I guess the `on-sync-status-unknown` and `on-sync-failed` triggers will need this check as well (and most likely `on-health-degraded` too).

My objection against solving it in the trigger condition itself is that anyone creating a user-defined trigger has to add this check as well (or perhaps not if they don't use the sync status in their template). This probably gets rather confusing. It sort of creates a relation between trigger condition and template, i.e. the condition has to know which variables are being used in the template. What is your opinion regarding this?

How about adding an [extra condition](https://github.com/argoproj-labs/argocd-notifications/blob/master/triggers/triggers.go#L153) evaluation here ? This condition is the above check and is defined by argocd-notifcations, but isn't part of 'every' trigger condition. Let me see if I can code something together to discuss further.

Reply:
interesting, the following condition works:
`time.Parse(app.status.operationState.finishedAt).After(time.Now())`
but this one does not:
`time.Parse(app.status.operationState.finishedAt).After(time.Parse(app.status.observedAt))`

`reflect: Call using zero Value argument` at the app in the second time.Parse

Reply:
the (built-in/user) defined trigger condition is wrapped with a sync status check:
https://github.com/argoproj-labs/argocd-notifications/pull/51/files#diff-56e03da81512d7963e3a61119bcbf36cR153

Reply:
I did not realize that check is required for all triggers. After thinking more about it I agree with you. I think we should just skip app reconciliation if sync operation just completed and reconciledAt is less than sync operation completion.